File: sdk/python/caribefiscal.py
"""
CaribeFiscal — SDK Python para la API e-CF de la DGII (República Dominicana).

Requiere `requests` (`pip install requests`). Compatible con Python 3.8+.

Ejemplo:
    from caribefiscal import CaribeFiscal
    cf = CaribeFiscal("https://ecf.omegaerp.do", "ck_live_xxx")
    res = cf.emit_invoice({...})
    print(res["eNCF"])
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional
import requests


class CaribeFiscalError(Exception):
    def __init__(self, message: str, status: int = 0, body: Any = None):
        super().__init__(message)
        self.status = status
        self.body = body


class CaribeFiscal:
    def __init__(self, base_url: str, api_key: str, timeout: int = 120):
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self._session = requests.Session()
        self._session.headers.update({"X-API-Key": api_key, "Accept": "application/json"})

    # ── e-CF ────────────────────────────────────────────────────────────────
    def emit_invoice(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        return self._req("POST", "/api/v1/ecf/submit", payload)

    def get_status(self, track_id: str) -> Dict[str, Any]:
        return self._req("GET", f"/api/v1/ecf/status/{track_id}")

    def get_invoice(self, encf: str) -> Dict[str, Any]:
        return self._req("GET", f"/api/v1/ecf/{encf}")

    def cancel_invoice(self, encf: str, motivo: str) -> Dict[str, Any]:
        return self._req("POST", "/api/v1/ecf/cancel", {"encf": encf, "motivo": motivo})

    def get_queue(self) -> Dict[str, Any]:
        return self._req("GET", "/api/v1/ecf/queue")

    def list_invoices(self, **filters: Any) -> Dict[str, Any]:
        return self._req("GET", "/api/v1/ecf/", params=filters)

    def list_inbound(self, **filters: Any) -> Dict[str, Any]:
        return self._req("GET", "/api/v1/inbound", params=filters)

    def download_pdf(self, encf: str) -> bytes:
        """Devuelve los bytes del PDF."""
        r = self._session.get(f"{self.base_url}/api/v1/ecf/{encf}/pdf", timeout=self.timeout)
        if r.status_code >= 400:
            raise CaribeFiscalError(f"HTTP {r.status_code}", r.status_code, r.content)
        return r.content

    # ── NCF ──────────────────────────────────────────────────────────────────
    def get_sequences(self) -> Dict[str, Any]:
        return self._req("GET", "/api/v1/ncf/sequences")

    def next_ncf(self, ecf_type: str) -> Dict[str, Any]:
        return self._req("GET", f"/api/v1/ncf/next/{ecf_type}")

    def configure_sequence(self, seq: Dict[str, Any]) -> Dict[str, Any]:
        return self._req("POST", "/api/v1/ncf/configure", seq)

    def deactivate_sequence(self, ecf_type: str) -> Dict[str, Any]:
        return self._req("DELETE", f"/api/v1/ncf/{ecf_type}")

    # ── RNC ──────────────────────────────────────────────────────────────────
    def lookup_rnc(self, rnc: str) -> Dict[str, Any]:
        return self._req("GET", f"/api/v1/rnc/{rnc}")

    def autocomplete_rnc(self, q: str) -> Dict[str, Any]:
        return self._req("GET", "/api/v1/rnc/autocomplete", params={"q": q})

    def bulk_rnc(self, rncs: List[str]) -> Dict[str, Any]:
        return self._req("POST", "/api/v1/rnc/bulk", {"rncs": rncs})

    # ── Reportes ──────────────────────────────────────────────────────────────
    def report_606(self, start: str, end: str) -> Dict[str, Any]:
        return self._report("606", start, end)

    def report_607(self, start: str, end: str) -> Dict[str, Any]:
        return self._report("607", start, end)

    def report_608(self, start: str, end: str) -> Dict[str, Any]:
        return self._report("608", start, end)

    def report_it1(self, start: str, end: str) -> Dict[str, Any]:
        return self._report("it1", start, end)

    def report_isr(self, start: str, end: str) -> Dict[str, Any]:
        return self._report("isr", start, end)

    def report_ir2(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._req("POST", "/api/v1/reports/ir2", body)

    def report_ir1(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._req("POST", "/api/v1/reports/ir1", body)

    def report_tss(self, periodo: str, empleados: List[Dict[str, Any]]) -> Dict[str, Any]:
        return self._req("POST", "/api/v1/reports/tss", {"periodo": periodo, "empleados": empleados})

    def _report(self, tipo: str, start: str, end: str) -> Dict[str, Any]:
        return self._req("POST", f"/api/v1/reports/{tipo}", {"startDate": start, "endDate": end})

    # ── Empresa ────────────────────────────────────────────────────────────────
    def get_config(self) -> Dict[str, Any]:
        return self._req("GET", "/api/v1/company/config")

    def update_config(self, cfg: Dict[str, Any]) -> Dict[str, Any]:
        return self._req("PUT", "/api/v1/company/config", cfg)

    def upload_certificate(self, p12_path: str, password: str) -> Dict[str, Any]:
        with open(p12_path, "rb") as fh:
            r = self._session.post(
                f"{self.base_url}/api/v1/company/certificate",
                params={"password": password},
                files={"file": ("cert.p12", fh, "application/x-pkcs12")},
                timeout=self.timeout,
            )
        return self._handle(r)

    # ── Núcleo HTTP ───────────────────────────────────────────────────────────
    def _req(self, method: str, path: str, body: Optional[Dict[str, Any]] = None,
             params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        params = {k: v for k, v in (params or {}).items() if v is not None and v != ""}
        r = self._session.request(
            method, f"{self.base_url}{path}",
            json=body if body is not None else None,
            params=params or None,
            timeout=self.timeout,
        )
        return self._handle(r)

    @staticmethod
    def _handle(r: requests.Response) -> Dict[str, Any]:
        try:
            data = r.json() if r.content else {}
        except ValueError:
            data = {}
        if r.status_code >= 400:
            msg = data.get("error") if isinstance(data, dict) else f"HTTP {r.status_code}"
            raise CaribeFiscalError(msg or f"HTTP {r.status_code}", r.status_code, data)
        return data

File: sdk/php/CaribeFiscal.php
<?php
/**
 * CaribeFiscal — SDK PHP para la API e-CF de la DGII (República Dominicana).
 *
 * Requiere la extensión cURL (estándar en PHP). Compatible con PHP 7.4+ / 8.x.
 *
 * Ejemplo:
 *   require 'CaribeFiscal.php';
 *   $cf = new CaribeFiscal('https://ecf.omegaerp.do', 'ck_live_xxx');
 *   $res = $cf->emitInvoice([...]);
 *   echo $res['eNCF'];
 */
class CaribeFiscalException extends \Exception
{
    /** @var int */    public $status;
    /** @var mixed */  public $body;
    public function __construct(string $message, int $status = 0, $body = null)
    {
        parent::__construct($message, $status);
        $this->status = $status;
        $this->body   = $body;
    }
}

class CaribeFiscal
{
    private string $baseUrl;
    private string $apiKey;
    private int    $timeout;

    public function __construct(string $baseUrl, string $apiKey, int $timeout = 120)
    {
        $this->baseUrl = rtrim($baseUrl, '/');
        $this->apiKey  = $apiKey;
        $this->timeout = $timeout;
    }

    // ── e-CF ────────────────────────────────────────────────────────────────
    public function emitInvoice(array $payload): array          { return $this->request('POST', '/api/v1/ecf/submit', $payload); }
    public function getStatus(string $trackId): array           { return $this->request('GET', "/api/v1/ecf/status/" . rawurlencode($trackId)); }
    public function getInvoice(string $encf): array             { return $this->request('GET', "/api/v1/ecf/" . rawurlencode($encf)); }
    public function cancelInvoice(string $encf, string $motivo): array { return $this->request('POST', '/api/v1/ecf/cancel', ['encf' => $encf, 'motivo' => $motivo]); }
    public function getQueue(): array                           { return $this->request('GET', '/api/v1/ecf/queue'); }
    public function listInvoices(array $filters = []): array    { return $this->request('GET', '/api/v1/ecf/' . $this->qs($filters)); }
    public function listInbound(array $filters = []): array     { return $this->request('GET', '/api/v1/inbound' . $this->qs($filters)); }

    /** Devuelve los bytes binarios del PDF. */
    public function downloadPdf(string $encf): string
    {
        return $this->request('GET', "/api/v1/ecf/" . rawurlencode($encf) . "/pdf", null, true);
    }

    // ── NCF ─────────────────────────────────────────────────────────────────
    public function getSequences(): array                       { return $this->request('GET', '/api/v1/ncf/sequences'); }
    public function nextNcf(string $type): array                { return $this->request('GET', "/api/v1/ncf/next/" . rawurlencode($type)); }
    public function configureSequence(array $seq): array        { return $this->request('POST', '/api/v1/ncf/configure', $seq); }
    public function deactivateSequence(string $type): array     { return $this->request('DELETE', "/api/v1/ncf/" . rawurlencode($type)); }

    // ── RNC ─────────────────────────────────────────────────────────────────
    public function lookupRnc(string $rnc): array              { return $this->request('GET', "/api/v1/rnc/" . rawurlencode($rnc)); }
    public function autocompleteRnc(string $q): array          { return $this->request('GET', '/api/v1/rnc/autocomplete' . $this->qs(['q' => $q])); }
    public function bulkRnc(array $rncs): array                { return $this->request('POST', '/api/v1/rnc/bulk', ['rncs' => $rncs]); }

    // ── Reportes ──────────────────────────────────────────────────────────────
    public function report606(string $start, string $end): array { return $this->report('606', $start, $end); }
    public function report607(string $start, string $end): array { return $this->report('607', $start, $end); }
    public function report608(string $start, string $end): array { return $this->report('608', $start, $end); }
    public function reportIt1(string $start, string $end): array { return $this->report('it1', $start, $end); }
    public function reportIsr(string $start, string $end): array { return $this->report('isr', $start, $end); }
    public function reportIr2(array $body): array               { return $this->request('POST', '/api/v1/reports/ir2', $body); }
    public function reportIr1(array $body): array               { return $this->request('POST', '/api/v1/reports/ir1', $body); }
    public function reportTss(string $periodo, array $empleados): array { return $this->request('POST', '/api/v1/reports/tss', ['periodo' => $periodo, 'empleados' => $empleados]); }

    private function report(string $tipo, string $start, string $end): array
    {
        return $this->request('POST', "/api/v1/reports/$tipo", ['startDate' => $start, 'endDate' => $end]);
    }

    // ── Empresa ───────────────────────────────────────────────────────────────
    public function getConfig(): array                         { return $this->request('GET', '/api/v1/company/config'); }
    public function updateConfig(array $cfg): array            { return $this->request('PUT', '/api/v1/company/config', $cfg); }

    /** Sube el certificado .p12. */
    public function uploadCertificate(string $p12Path, string $password): array
    {
        $url = $this->baseUrl . '/api/v1/company/certificate?password=' . rawurlencode($password);
        $ch  = curl_init($url);
        curl_setopt_array($ch, [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT        => $this->timeout,
            CURLOPT_POST           => true,
            CURLOPT_HTTPHEADER     => ['X-API-Key: ' . $this->apiKey],
            CURLOPT_POSTFIELDS     => ['file' => new \CURLFile($p12Path, 'application/x-pkcs12', basename($p12Path))],
        ]);
        return $this->exec($ch, false);
    }

    // ── Núcleo HTTP ─────────────────────────────────────────────────────────
    private function qs(array $params): string
    {
        $params = array_filter($params, fn($v) => $v !== null && $v !== '');
        return $params ? '?' . http_build_query($params) : '';
    }

    private function request(string $method, string $path, $body = null, bool $raw = false)
    {
        $ch = curl_init($this->baseUrl . $path);
        $headers = ['X-API-Key: ' . $this->apiKey, 'Accept: application/json'];
        $opts = [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT        => $this->timeout,
            CURLOPT_CUSTOMREQUEST  => $method,
        ];
        if ($body !== null) {
            $opts[CURLOPT_POSTFIELDS] = json_encode($body);
            $headers[] = 'Content-Type: application/json';
        }
        $opts[CURLOPT_HTTPHEADER] = $headers;
        curl_setopt_array($ch, $opts);
        return $this->exec($ch, $raw);
    }

    private function exec($ch, bool $raw)
    {
        $response = curl_exec($ch);
        if ($response === false) {
            $err = curl_error($ch);
            curl_close($ch);
            throw new CaribeFiscalException("Error de red: $err");
        }
        $status = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        if ($raw) {
            if ($status >= 400) throw new CaribeFiscalException("HTTP $status", $status, $response);
            return $response;
        }

        $data = json_decode($response, true);
        if ($status >= 400) {
            $msg = is_array($data) && isset($data['error']) ? $data['error'] : "HTTP $status";
            throw new CaribeFiscalException($msg, $status, $data);
        }
        return $data ?? [];
    }
}
